File: apps/minecraft/spigot/src/main/java/crabcraft/net/crabUtilities/jade/protocol/provider/block/BeehiveProvider.java
package crabcraft.net.crabUtilities.jade.protocol.provider.block;

import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.resources.Identifier;
import net.minecraft.world.level.block.entity.BeehiveBlockEntity;
import org.jetbrains.annotations.NotNull;
import crabcraft.net.crabUtilities.jade.protocol.JadeProtocol;
import crabcraft.net.crabUtilities.jade.protocol.accessor.BlockAccessor;
import crabcraft.net.crabUtilities.jade.protocol.provider.StreamServerDataProvider;

public enum BeehiveProvider implements StreamServerDataProvider<BlockAccessor, Byte> {
    INSTANCE;

    private static final Identifier MC_BEEHIVE = JadeProtocol.mc_id("beehive");

    @Override
    public @NotNull StreamCodec<RegistryFriendlyByteBuf, Byte> streamCodec() {
        return ByteBufCodecs.BYTE.cast();
    }

    @Override
    public Byte streamData(@NotNull BlockAccessor accessor) {
        BeehiveBlockEntity beehive = (BeehiveBlockEntity) accessor.getBlockEntity();
        int bees = beehive.getOccupantCount();
        return (byte) (beehive.isFull() ? bees : -bees);
    }

    @Override
    public Identifier getUid() {
        return MC_BEEHIVE;
    }
}

File: apps/minecraft/spigot/src/main/kotlin/crabcraft/net/crabUtilities/jade/protocol/provider/block/BeehiveProvider.kt
package crabcraft.net.crabUtilities.jade.protocol.provider.block

import crabcraft.net.crabUtilities.jade.protocol.JadeProtocol
import crabcraft.net.crabUtilities.jade.protocol.accessor.BlockAccessor
import crabcraft.net.crabUtilities.jade.protocol.provider.StreamServerDataProvider
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.world.level.block.entity.BeehiveBlockEntity

enum class BeehiveProvider : StreamServerDataProvider<BlockAccessor, Byte> {
    INSTANCE;

    override fun streamData(accessor: BlockAccessor): Byte? {
        val beehive = accessor.getBlockEntity() as BeehiveBlockEntity
        val bees = beehive.getOccupantCount()
        return (if (beehive.isFull()) bees else -bees).toByte()
    }

    override fun streamCodec(): StreamCodec<RegistryFriendlyByteBuf, Byte> = ByteBufCodecs.BYTE.cast()

    override fun getUid() = UID

    companion object {
        private val UID = JadeProtocol.mc_id("beehive")
    }
}
